ledger-core/insolar/payload/validation_vstatereport.go

```diff
@@ -6,51 +6,77 @@
 package payload
 
 import (
+	"github.com/insolar/assured-ledger/ledger-core/pulse"
 	"github.com/insolar/assured-ledger/ledger-core/vanilla/throw"
 )
 
 var _ Validatable = &VStateReport{}
 
-func (m *VStateReport) Validate(currPulse PulseNumber) error {
+type vStateReportValidateStatusFunc func(pulse.Number, pulse.Number) error
+
+func (m *VStateReport) Validate(currentPulse pulse.Number) error {
 	if err := m.validateUnimplemented(); err != nil {
 		return err
 	}
 
-	if asOf := m.GetAsOf(); !asOf.IsTimePulse() || asOf >= currPulse {
+	if !isTimePulseBefore(m.GetAsOf(), currentPulse) {
 		return throw.New("AsOf should be time pulse and less that current pulse")
 	}
 
-	switch object := m.GetObject(); {
-	case !object.IsSelfScope():
-		return throw.New("Object reference should be self scoped")
-	case !object.GetLocal().Pulse().IsTimePulse():
-		fallthrough
-	case object.GetLocal().Pulse() >= currPulse:
-		return throw.New("Object pulse should be valid time pulse less that current pulse")
+	objectPulse, err := validSelfScopedGlobalWithPulseBeforeOrEq(m.Object, currentPulse, "Object")
+	if err != nil {
+		return err
+	}
+
+	validateStatus := map[VStateReport_StateStatus]vStateReportValidateStatusFunc{
```

Comment:
@bigbes why is it better to use map instead switch/case?

Reply:
Unification of check signatures, lesser code duplication; no other practical reason
switch-case may be still faster, can be rewritten as `func(VStateReport_StateStatus) vStateReportValidateStatusFunc` with switch case for functions (combining that two decisions)

Reply:
What is the duplication code? We had 4 methods and switch/case. now we have an additional type for function + map and the same 4 methods.

Reply:
every `case` have exactly same code with only one difference: callback

Reply:
This code example the same simple, but no need a new type and map.
```
var err error

	switch m.GetStatus() {
	case Ready:
		err = m.validateStatusReady(objectPulse, currentPulse)
	case Empty:
		err = m.validateStatusEmpty(objectPulse, currentPulse)
	case Missing, Inactive:
		err = m.validateStatusMissingOrInactive()
	default:
		return throw.New("Unexpected state received")
	}

	if err != nil {
		return err
	}
```

Reply:
fixed